lib/queryContextCache.js

```diff
@@ -0,0 +1,391 @@
+/*
+ * Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
+ */
+const Logger = require('./logger');
+const crypto = require('crypto');
+/**
+     * Constructor.
+     *
+     * @param id database id
+     * @param timestamp Server time when this entry read
+     * @param priority Priority of this entry w.r.t other ids
+     * @param context Opaque query context, used by query processor in the server.
+     */
+
+/** Query context information. */
+function QueryContextElement (id,timestamp,priority,context) {
+  this.id = id;
+  this.timestamp = timestamp;
+  this.priority = priority;
+  this.context = context;
+}
+
+QueryContextElement.prototype.getId = function () {
+  return this.id;
+};
+
+QueryContextElement.prototype.getReadTimestamp = function () {
+  return this.timestamp;
+};
+
+QueryContextElement.prototype.getPriority = function () {
+  return this.priority;
+};
+
+QueryContextElement.prototype.getContext = function () {
+  return this.context;
+};
+ 
+QueryContextElement.prototype.setId = function (id) {
+  this.id = id;
+};
+
+QueryContextElement.prototype.setReadTimestamp = function (timestamp) {
+  this.timestamp = timestamp;
+};
+
+QueryContextElement.prototype.setPriority = function (priority) {
+  this.priority = priority;
+};
+
+QueryContextElement.prototype.setContext = function (context) {
+  this.context = context;
+};
+
+QueryContextElement.prototype.equals = function (obj) {
+  if (!(obj instanceof QueryContextElement)) {
+    return false;
+  }
+    
+  return (this.id === obj.id
+        && this.timestamp === obj.timestamp
+        && this.priority === obj.priority
+        && this.context.equals(obj.context));
+};
+
+QueryContextElement.prototype.hashCode = function () {
+  let hash = 31;
+  hash = hash * 31 + parseInt(this.id);
+  hash += (hash * 31) + parseInt(this.timestamp);
+  hash += (hash * 31) + parseInt(this.priority);
+  hash += (hash * 31) + parseInt(crypto.createHash('md5').update(Buffer.from(this.context,'utf-8').toString()).digest(),2);
+
+  return hash;
+};
+
+/**
+ * Most Recently Used and Priority based cache. A separate cache for each connection in the driver.
+ */
+
+/**
+   * Constructor.
+   *
+   * @param capacity Maximum capacity of the cache.
+   */
+
+function QueryContextCache (capacity) {
+  this.capacity = capacity;
+  this.idMap = new Map(); // Map for id and QCC
+  this.treeSet = new Set(); // Order data as per priority
+  this.priorityMap = new Map(); // Map for priority and QCC
+  this.newPriorityMap = new Map(); // Intermediate map for priority and QCC for current round of merging
```

Comment:
when it's intermediate - should it be a field?

Reply:
about the proces - why do we have to process query context elements one by one? we receive the list of query context elements in the response so let's merge them once and don't keep intermediate results

Reply:
Removed, and directly insert elements into priorityMap